File: gateway/app/schemas/code.py
"""Pydantic schemas for Code Assistant endpoints -- Phase 5."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Workspace schemas
# ---------------------------------------------------------------------------

class WorkspaceCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str = ""


class WorkspaceOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str
    path: str
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    file_tree: "list[FileTreeNode] | None" = None

    model_config = {"from_attributes": True}


class WorkspaceListOut(BaseModel):
    workspaces: list[WorkspaceOut]
    total: int


# ---------------------------------------------------------------------------
# File tree schemas
# ---------------------------------------------------------------------------

class FileTreeNode(BaseModel):
    name: str
    path: str  # relative path within workspace
    is_dir: bool = False
    size: int = 0
    language: str | None = None
    line_count: int | None = None
    children: "list[FileTreeNode] | None" = None


class FileContent(BaseModel):
    path: str
    content: str
    language: str | None = None
    size: int = 0
    line_count: int = 0


class FileWriteRequest(BaseModel):
    content: str


# ---------------------------------------------------------------------------
# Code execution schemas
# ---------------------------------------------------------------------------

class ExecuteRequest(BaseModel):
    code: str = Field(..., min_length=1)
    language: str = Field(default="python", pattern=r"^(python|javascript|bash|sql)$")
    session_id: uuid.UUID | None = None
    timeout: int = Field(default=30, ge=1, le=300)


class ExecuteResponse(BaseModel):
    execution_id: uuid.UUID | None = None
    session_id: uuid.UUID | None = None
    stdout: str = ""
    stderr: str = ""
    return_value: str | None = None
    exit_code: int = 0
    execution_time_ms: float = 0.0
    language: str = "python"


# ---------------------------------------------------------------------------
# Code analysis schemas
# ---------------------------------------------------------------------------

class AnalysisIssue(BaseModel):
    line: int | None = None
    column: int | None = None
    severity: str = "warning"  # info, warning, error
    message: str = ""
    rule: str | None = None


class AnalyzeRequest(BaseModel):
    code: str = Field(..., min_length=1)
    language: str = "python"
    analysis_type: str = Field(
        default="full",
        description="Type of analysis: full, security, bugs, style",
    )


class AnalyzeResponse(BaseModel):
    issues: list[AnalysisIssue] = Field(default_factory=list)
    summary: str = ""
    issue_count: int = 0
    language: str = "python"


# ---------------------------------------------------------------------------
# Code generation / explanation / review schemas
# ---------------------------------------------------------------------------

class CodeGenRequest(BaseModel):
    prompt: str = Field(..., min_length=1, max_length=16000)
    language: str = "python"
    context: str | None = None


class CodeGenResponse(BaseModel):
    code: str = ""
    explanation: str = ""
    language: str = "python"


class ExplainRequest(BaseModel):
    code: str = Field(..., min_length=1)
    language: str = "python"
    detail_level: str = Field(
        default="normal",
        description="Level of detail: brief, normal, detailed",
    )


class ExplainResponse(BaseModel):
    explanation: str = ""
    language: str = "python"


class ReviewRequest(BaseModel):
    code: str | None = None
    diff: str | None = None
    language: str = "python"
    focus: str = Field(
        default="general",
        description="Review focus: general, security, performance, bugs",
    )


class ReviewFinding(BaseModel):
    line: int | None = None
    severity: str = "info"  # info, warning, error, critical
    category: str = ""
    message: str = ""
    suggestion: str | None = None


class ReviewResponse(BaseModel):
    findings: list[ReviewFinding] = Field(default_factory=list)
    summary: str = ""
    overall_quality: str = "acceptable"  # poor, acceptable, good, excellent


# ---------------------------------------------------------------------------
# Git schemas
# ---------------------------------------------------------------------------

class GitDiffSummaryRequest(BaseModel):
    diff: str = Field(..., min_length=1)


class GitDiffSummary(BaseModel):
    summary: str = ""
    files_changed: int = 0
    additions: int = 0
    deletions: int = 0
    file_summaries: list[dict[str, str]] = Field(default_factory=list)


class CommitMessageRequest(BaseModel):
    diff: str = Field(..., min_length=1)
    style: str = Field(
        default="conventional",
        description="Commit message style: conventional, descriptive, brief",
    )


class CommitMessageResponse(BaseModel):
    message: str = ""
    subject: str = ""
    body: str = ""


# Resolve forward references
WorkspaceOut.model_rebuild()
FileTreeNode.model_rebuild()

File: frontend/src/lib/code.ts
/**
 * Code Assistant API helpers -- Phase 5.
 */

import { apiJson, apiFetch } from "./api";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Workspace {
  id: string;
  name: string;
  description: string;
  path: string;
  user_id: string;
  created_at: string;
  updated_at: string;
  file_tree?: FileTreeNode[];
}

export interface FileTreeNode {
  name: string;
  path: string;
  is_dir: boolean;
  size: number;
  language: string | null;
  line_count: number | null;
  children: FileTreeNode[] | null;
}

export interface FileContent {
  path: string;
  content: string;
  language: string | null;
  size: number;
  line_count: number;
}

export interface ExecuteResult {
  execution_id: string | null;
  session_id: string | null;
  stdout: string;
  stderr: string;
  return_value: string | null;
  exit_code: number;
  execution_time_ms: number;
  language: string;
}

export interface AnalysisIssue {
  line: number | null;
  column: number | null;
  severity: string;
  message: string;
  rule: string | null;
}

export interface AnalyzeResult {
  issues: AnalysisIssue[];
  summary: string;
  issue_count: number;
  language: string;
}

export interface CodeGenResult {
  code: string;
  explanation: string;
  language: string;
}

export interface ExplainResult {
  explanation: string;
  language: string;
}

export interface ReviewFinding {
  line: number | null;
  severity: string;
  category: string;
  message: string;
  suggestion: string | null;
}

export interface ReviewResult {
  findings: ReviewFinding[];
  summary: string;
  overall_quality: string;
}

export interface DiffSummary {
  summary: string;
  files_changed: number;
  additions: number;
  deletions: number;
  file_summaries: { file: string; changes: string }[];
}

export interface CommitMessage {
  message: string;
  subject: string;
  body: string;
}

// ---------------------------------------------------------------------------
// Workspace API
// ---------------------------------------------------------------------------

export async function createWorkspace(
  name: string,
  description: string = ""
): Promise<Workspace> {
  return apiJson<Workspace>("/api/code/workspaces", {
    method: "POST",
    body: JSON.stringify({ name, description }),
  });
}

export async function listWorkspaces(): Promise<{
  workspaces: Workspace[];
  total: number;
}> {
  return apiJson("/api/code/workspaces");
}

export async function getWorkspace(wsId: string): Promise<Workspace> {
  return apiJson<Workspace>(`/api/code/workspaces/${wsId}`);
}

export async function deleteWorkspace(wsId: string): Promise<void> {
  await apiFetch(`/api/code/workspaces/${wsId}`, { method: "DELETE" });
}

export async function readFile(
  wsId: string,
  path: string
): Promise<FileContent> {
  return apiJson<FileContent>(
    `/api/code/workspaces/${wsId}/files/${path}`
  );
}

export async function writeFile(
  wsId: string,
  path: string,
  content: string
): Promise<FileContent> {
  return apiJson<FileContent>(
    `/api/code/workspaces/${wsId}/files/${path}`,
    {
      method: "PUT",
      body: JSON.stringify({ content }),
    }
  );
}

export async function deleteFile(
  wsId: string,
  path: string
): Promise<void> {
  await apiFetch(`/api/code/workspaces/${wsId}/files/${path}`, {
    method: "DELETE",
  });
}

// ---------------------------------------------------------------------------
// Code execution API
// ---------------------------------------------------------------------------

export async function executeCode(
  code: string,
  language: string = "python",
  sessionId?: string,
  timeout: number = 30
): Promise<ExecuteResult> {
  return apiJson<ExecuteResult>("/api/code/execute", {
    method: "POST",
    body: JSON.stringify({
      code,
      language,
      session_id: sessionId || null,
      timeout,
    }),
  });
}

// ---------------------------------------------------------------------------
// Code analysis API
// ---------------------------------------------------------------------------

export async function analyzeCode(
  code: string,
  language: string = "python",
  analysisType: string = "full"
): Promise<AnalyzeResult> {
  return apiJson<AnalyzeResult>("/api/code/analyze", {
    method: "POST",
    body: JSON.stringify({ code, language, analysis_type: analysisType }),
  });
}

export async function explainCode(
  code: string,
  language: string = "python",
  detailLevel: string = "normal"
): Promise<ExplainResult> {
  return apiJson<ExplainResult>("/api/code/explain", {
    method: "POST",
    body: JSON.stringify({ code, language, detail_level: detailLevel }),
  });
}

export async function generateCode(
  prompt: string,
  language: string = "python",
  context?: string
): Promise<CodeGenResult> {
  return apiJson<CodeGenResult>("/api/code/generate", {
    method: "POST",
    body: JSON.stringify({ prompt, language, context: context || null }),
  });
}

export async function reviewCode(
  code?: string,
  diff?: string,
  language: string = "python",
  focus: string = "general"
): Promise<ReviewResult> {
  return apiJson<ReviewResult>("/api/code/review", {
    method: "POST",
    body: JSON.stringify({
      code: code || null,
      diff: diff || null,
      language,
      focus,
    }),
  });
}

// ---------------------------------------------------------------------------
// Git API
// ---------------------------------------------------------------------------

export async function getDiffSummary(diff: string): Promise<DiffSummary> {
  return apiJson<DiffSummary>("/api/code/git/diff-summary", {
    method: "POST",
    body: JSON.stringify({ diff }),
  });
}

export async function getCommitMessage(
  diff: string,
  style: string = "conventional"
): Promise<CommitMessage> {
  return apiJson<CommitMessage>("/api/code/git/commit-message", {
    method: "POST",
    body: JSON.stringify({ diff, style }),
  });
}
